### DISCUSSION TITLE
integrations/kaggle/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#17971] [Docs]
# integrations/kaggle/

Dive into our guide on YOLO11's integration with Kaggle. Find out what Kaggle is, its key features, and how to train a YOLO11 model using the integration.

https://docs.ultralytics.com/integrations/kaggle/

**Comment by UltralyticsAssistant** [comment]
👋 Hello, thank you for bringing up YOLO’s integration with Kaggle 🚀! It’s exciting to see interest in leveraging Kaggle for YOLO training and experiments. For in-depth guidance, we recommend reviewing the [Kaggle Integration Documentation](https://docs.ultralytics.com/integrations/kaggle/) where you’ll find detailed instructions on how to get started.

If you’re encountering any issues or have a ❓ Question, we’d love to help! Please provide as much information as possible, such as error logs, your training setup, or dataset details. If this is a 🐛 Bug Report, kindly include a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to enable us to diagnose the issue efficiently.

## Getting Started with Kaggle

Train YOLO models directly on Kaggle with just a few steps. You can access pre-configured Kaggle notebooks here: <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>

Don’t forget to ensure your environment is up to date:

```bash
pip install -U ultralytics
```

For comprehensive instructions on YOLO environments, integrations, and supported platforms, check out the [Docs](https://docs.ultralytics.com).

## Join the Ultralytics Community

Engage with other YOLO enthusiasts and developers in the vibrant Ultralytics community! Here’s how:
- Real-time discussions on [Discord](https://discord.com/invite/ultralytics) 🎧
- Longer, detailed posts on [Discourse](https://community.ultralytics.com)
- Fun and knowledge-sharing on our [Subreddit](https://reddit.com/r/Ultralytics) 💡

## Stay Updated

Ensure your setup is aligned with the latest updates and features by using a verified environment:
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM: See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI: See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker** Images: See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

Our team of engineers will check in soon to provide further assistance if needed. Thank you for exploring YOLO with Ultralytics! 🚀

## CI Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

Keep an eye on the project's CI status for the latest updates and stability assessments. Green means good to go ✅!

**Comment by md-jawad-117** [comment]
Hello, I was using the GPU T4 x2 for training my custom RT-DETR model which is infact was twice as fast compared to P100 GPU with just one of them running.🚀🚀 While the second one was idle.                                                                                                                                                                                         Now when I tried to utilize both GPUs using  device=[0, 1] & device=[-1, -1], i received an error which persisted as long as I did not restart the kernel. ⛓️‍💥⛓️‍💥

from ultralytics import RTDETR

model = RTDETR('/kaggle/working/rtdr.yaml').load('rtdetr-l.pt')

model.info()

results = model.train(data="/kaggle/working/data.yaml", epochs=30, imgsz=512, optimizer='AdamW', plots=True, batch=32, lr0=0.0001, device=[0,1])

**Reply by md-jawad-117** [reply]
Error Msg: 
Transferred 725/725 items from pretrained weights
DDP: debug command /usr/bin/python3 -m torch.distributed.run --nproc_per_node 2 --master_port 52415 /root/.config/Ultralytics/DDP/_temp_xi8731gh140405119848080.py
Traceback (most recent call last):
  File "/root/.config/Ultralytics/DDP/_temp_xi8731gh140405119848080.py", line 6, in <module>
    from ultralytics.models.rtdetr.train import RTDETRTrainer
ModuleNotFoundError: No module named 'ultralytics'
Traceback (most recent call last):
  File "/root/.config/Ultralytics/DDP/_temp_xi8731gh140405119848080.py", line 6, in <module>
    from ultralytics.models.rtdetr.train import RTDETRTrainer
ModuleNotFoundError: No module named 'ultralytics'
E0601 09:54:34.472000 132 torch/distributed/elastic/multiprocessing/api.py:869] failed (exitcode: 1) local_rank: 0 (pid: 134) of binary: /usr/bin/python3
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/usr/local/lib/python3.11/dist-packages/torch/distributed/run.py", line 922, in <module>
    main()
  File "/usr/local/lib/python3.11/dist-packages/torch/distributed/elastic/multiprocessing/errors/__init__.py", line 355, in wrapper
    return f(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/torch/distributed/run.py", line 918, in main
    run(args)
  File "/usr/local/lib/python3.11/dist-packages/torch/distributed/run.py", line 909, in run
    elastic_launch(
  File "/usr/local/lib/python3.11/dist-packages/torch/distributed/launcher/api.py", line 138, in __call__
    return launch_agent(self._config, self._entrypoint, list(args))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/torch/distributed/launcher/api.py", line 269, in launch_agent
    raise ChildFailedError(
torch.distributed.elastic.multiprocessing.errors.ChildFailedError: 
============================================================
/root/.config/Ultralytics/DDP/_temp_xi8731gh140405119848080.py FAILED
------------------------------------------------------------
Failures:
[1]:
  time      : 2025-06-01_09:54:34
  host      : 9bca5bb85783
  rank      : 1 (local_rank: 1)
  exitcode  : 1 (pid: 135)
  error_file: <N/A>
  traceback : To enable traceback see: https://pytorch.org/docs/stable/elastic/errors.html
------------------------------------------------------------
Root Cause (first observed failure):
[0]:
  time      : 2025-06-01_09:54:34
  host      : 9bca5bb85783
  rank      : 0 (local_rank: 0)
  exitcode  : 1 (pid: 134)
  error_file: <N/A>
  traceback : To enable traceback see: https://pytorch.org/docs/stable/elastic/errors.html
============================================================
---------------------------------------------------------------------------
CalledProcessError                        Traceback (most recent call last)
/tmp/ipykernel_35/2411087223.py in <cell line: 0>()
----> 1 results = model.train(data="/kaggle/working/data.yaml", epochs=30, imgsz=512, optimizer='AdamW', plots=True, batch=16, lr0=0.0001, device=[0,-1])

/kaggle/working/ultralytics/engine/model.py in train(self, trainer, **kwargs)
    795 
    796         self.trainer.hub_session = self.session  # attach optional HUB session
--> 797         self.trainer.train()
    798         # Update model and cfg after training
    799         if RANK in {-1, 0}:

/kaggle/working/ultralytics/engine/trainer.py in train(self)
    220                 subprocess.run(cmd, check=True)
    221             except Exception as e:
--> 222                 raise e
    223             finally:
    224                 ddp_cleanup(self, str(file))

/kaggle/working/ultralytics/engine/trainer.py in train(self)
    218             try:
    219                 LOGGER.info(f"{colorstr('DDP:')} debug command {' '.join(cmd)}")
--> 220                 subprocess.run(cmd, check=True)
    221             except Exception as e:
    222                 raise e

/usr/lib/python3.11/subprocess.py in run(input, capture_output, timeout, check, *popenargs, **kwargs)
    569         retcode = process.poll()
    570         if check and retcode:
--> 571             raise CalledProcessError(retcode, process.args,
    572                                      output=stdout, stderr=stderr)
    573     return CompletedProcess(process.args, retcode, stdout, stderr)

CalledProcessError: Command '['/usr/bin/python3', '-m', 'torch.distributed.run', '--nproc_per_node', '2', '--master_port', '52415', '/root/.config/Ultralytics/DDP/_temp_xi8731gh140405119848080.py']' returned non-zero exit status 1.

**Reply by glenn-jocher** [reply]
Thanks for sharing your multi-GPU training issue with RT-DETR on Kaggle! The error you're encountering is related to DDP (Distributed Data Parallel) not being able to find the Ultralytics module in the spawned processes.

This is a known issue on Kaggle when using multi-GPU training with DDP. The spawned processes don't inherit the current environment properly. Here are a few solutions to try:

1. **Install Ultralytics system-wide** before training:
```python
import subprocess
subprocess.run(['pip', 'install', '-U', 'ultralytics'], check=True)
```

2. **Use single GPU with larger batch size** instead:
```python
results = model.train(data="/kaggle/working/data.yaml", epochs=30, imgsz=512, 
                     optimizer='AdamW', plots=True, batch=64, lr0=0.0001, device=0)
```

3. **Try using device='cpu'** to verify your setup works, then switch back to single GPU.

Unfortunately, Kaggle's multi-GPU DDP setup can be unreliable due to environment isolation issues. Since you mentioned T4 x2 was already twice as fast as P100 with just one GPU active, you might get better results focusing on optimizing single-GPU training with larger batch sizes or mixed precision training.

Have you tried increasing the batch size on a single T4 GPU to maximize utilization?
